platform/api/esx/api.go

```diff
@@ -0,0 +1,742 @@
+	}
+
+	vm, err := defaults.finder.VirtualMachine(a.ctx, vmName)
+	if err != nil {
+		return fmt.Errorf("couldn't find VM: %v", err)
+	}
+
+	return a.deleteDevice(vm)
+}
+
+func (a *API) deleteDevice(vm *object.VirtualMachine) error {
+	task, err := vm.PowerOff(a.ctx)
+	if err != nil {
+		return fmt.Errorf("powering off vm: %v", err)
+	}
+
+	// We don't check for errors on this task because it will throw an error
+	// if the VM is already in a powered off state
+	_ = task.Wait(a.ctx)
+
+	task, err = vm.Destroy(a.ctx)
+	if err != nil {
+		return fmt.Errorf("destroying vm: %v", vm)
+	}
+
+	return task.Wait(a.ctx)
+}
+
+func (a *API) buildCISR(name string, flagOVAPath string, finder *find.Finder, defaultNetwork object.NetworkReference, resourcePool *object.ResourcePool, datastore *object.Datastore) (*Archive, *types.OvfCreateImportSpecResult, error) {
+	var nets []types.OvfNetworkMapping
+	nets = append(nets, types.OvfNetworkMapping{
+		Name:    "test",
+		Network: defaultNetwork.Reference(),
+	})
+
+	archive := &Archive{flagOVAPath}
+	envelope, err := archive.ReadEnvelope("*.ovf")
+	if err != nil {
+		return nil, nil, fmt.Errorf("reading envelope: %v", err)
+	}
+
+	ovfHandler := object.NewOvfManager(a.client.Client)
+	cisp := types.OvfCreateImportSpecParams{
+		EntityName: name,
+		OvfManagerCommonParams: types.OvfManagerCommonParams{
+			Locale: "US"},
+		PropertyMapping: []types.KeyValue{},
+		NetworkMapping:  networkMap(finder, envelope),
+	}
+
+	descriptor, err := archive.ReadOvf("*.ovf")
+	if err != nil {
+		return nil, nil, fmt.Errorf("reading ovf: %v", err)
+	}
+	cisr, err := ovfHandler.CreateImportSpec(a.ctx, string(descriptor), resourcePool, datastore, cisp)
+	if err != nil {
+		return nil, nil, err
+	}
+	if cisr.Error != nil {
+		return nil, nil, errors.New(cisr.Error[0].LocalizedMessage)
+	}
+	return archive, cisr, nil
+}
+
+func (a *API) getServerDefaults() (ServerResources, error) {
+	finder := find.NewFinder(a.client.Client, true)
+	datacenter, err := finder.DefaultDatacenter(a.ctx)
+	if err != nil {
+		return ServerResources{}, err
+	}
+	finder.SetDatacenter(datacenter)
+	resourcePool, err := finder.DefaultResourcePool(a.ctx)
+	if err != nil {
+		return ServerResources{}, err
+	}
+	datastore, err := finder.DefaultDatastore(a.ctx)
+	if err != nil {
+		return ServerResources{}, err
+	}
+
+	defaultNetwork, err := finder.DefaultNetwork(a.ctx)
+	if err != nil {
+		return ServerResources{}, err
+	}
+
+	return ServerResources{
+		finder:       finder,
+		datacenter:   datacenter,
+		resourcePool: resourcePool,
+		datastore:    datastore,
+		network:      defaultNetwork,
+	}, nil
+}
+
+func (a *API) uploadToResourcePool(archive *Archive, resourcePool *object.ResourcePool, cisr *types.OvfCreateImportSpecResult, folder *object.Folder) (*types.ManagedObjectReference, error) {
+	lease, err := resourcePool.ImportVApp(a.ctx, cisr.ImportSpec, folder, nil)
+	if err != nil {
+		return nil, err
```

Comment:
When to wrap errors is a judgment call, but please `err` on the side of wrapping if there's any chance the context could be ambiguous.